crates/orderbook/src/orderbook.rs

```diff
@@ -91,13 +92,29 @@ pub enum AddOrderError {
     OrderValidation(ValidationError),
     #[error("database error: {0}")]
     Database(#[from] sqlx::Error),
+    #[error(
+        "contract app data {contract_app_data:?} is associated with full app data {existing:?} \
+         which is different from the provided {provided:?}"
+    )]
+    AppDataMismatch {
+        contract_app_data: AppDataHash,
+        provided: String,
+        existing: String,
+    },
 }
 
-impl From<InsertionError> for AddOrderError {
-    fn from(err: InsertionError) -> Self {
+impl AddOrderError {
+    fn from_insertion(err: InsertionError, order: &Order) -> Self {
         match err {
             InsertionError::DuplicatedRecord => AddOrderError::DuplicatedOrder,
             InsertionError::DbError(err) => AddOrderError::Database(err),
+            InsertionError::AppDataMismatch(existing) => AddOrderError::AppDataMismatch {
+                contract_app_data: order.data.app_data,
+                // Unwrap because this error can only occur if full app data was set.
+                provided: order.metadata.full_app_data.clone().unwrap(),
+                // Unwrap because we only store utf-8 full app data.
+                existing: String::from_utf8(existing).unwrap(),
```

Comment:
I wouldn't unwrap here, as this assumes a specific format for `app_data` which is technically "outside the scope of this module". You can just use `String::from_utf8_lossy` to make this "fail-proof".

Reply:
The problem with that is that we could get silent changes to app data in the conversion which would later confuse us. I could add an error log but not sure if that is better than unwrap for something that shouldn't happen.

Reply:
> The problem with that is that we could get silent changes to app data in the conversion which would later confuse us

This is just error context that we return in the API right? If anything I find this makes encoding issues more visible?

Reply:
Went with the log.